app/models/user.rb

```diff
@@ -0,0 +1,42 @@
+class User < ApplicationRecord
+  has_many :products # add scope so just logged in user has products
+  has_many :order_items, through: :products
+  has_many :orders, through: :order_items
+  validates :username, uniqueness: true, presence: true
+  validates :uid, uniqueness: {scope: :provider}, presence: true
+  validates :email, :provider, presence: :true
+
+  def self.build_from_github(auth_hash)
+    user = User.new
+    user.uid = auth_hash["uid"]
+    user.provider = 'github'
+    user.name = auth_hash["info"]["name"]
+    user.username = auth_hash["info"]["name"]
```

Comment:
Line 14 causes the login bug that I mentioned in the feedback notes. In the auth_hash ["info"]["name"] is an optional field and can be nil. If it is nil (as mine was when I tested) this causes the validation that the username has to exist to fail. To fix this, I would suggest using auth_hash["info"]["login"] as an alternative to name, or removing the validation. It is important to check that if you are validating on something, it will always exist in a valid case.